### PR TITLE
Add a GHA to check PRs on 4.0, blocking them if label is not present (only post release date)

### PR DESCRIPTION
**Pull request by jhchabran**
Adds a GitHub action that will check that after *2022-09-09* the `must-be-in-4.0` label is present, or the check will fail otherwise. If we are before that day, the check will always be green.

Later this week, I'll add this check as mandatory.

## Test plan

- Run with dday set to 2021-09-09 and no label:
  - 🔴 https://github.com/sourcegraph/sourcegraph/runs/8072123819?check_suite_focus=true
- Run with dday set to 2021-09-09 and with the label: 
  - 🟢 https://github.com/sourcegraph/sourcegraph/runs/8072108225?check_suite_focus=true
- Run with dday set to *2022-09-09* and no label: 
  - 🟢 https://github.com/sourcegraph/sourcegraph/runs/8072144421?check_suite_focus=true

### CODE REVIEW

**Comment by jhchabran** [comment]
cc @keegancsmith as this was your idea 😊

File: .github/workflows/fourpointoh-protector.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+name: Release 4.0 protection
```
Comment (mrnugget) [inline .github/workflows/fourpointoh-protector.yml:1]:
```suggestion
name: Sourcegraph 4.0 code freeze
```

File: .github/workflows/fourpointoh-protector.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      - main
+
+jobs:
+  protect-pr:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check date and labels
+        id: check-date-and-labels
+        run: |
+          has_label="${{contains(github.event.pull_request.labels.*.name, 'must-go-in-4.0')}}"
```
Comment (mrnugget) [inline .github/workflows/fourpointoh-protector.yml:16]:
```suggestion
          has_label="${{contains(github.event.pull_request.labels.*.name, 'i-acknowledge-this-goes-into-4.0')}}"
```

File: .github/workflows/fourpointoh-protector.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+          today=$(date +'%Y%m%d%H%M')
+          dday="202209090000"
+
+          if [ "$today" -gt "$dday" ]; then
+            if [ "$has_label" = "true" ]; then
+              echo "✅ Label 'must-go-in-4.0' is present"
```
Comment (mrnugget) [inline .github/workflows/fourpointoh-protector.yml:22]:
```suggestion
              echo "✅ Label 'i-acknowledge-this-goes-into-4.0' is present"
```

File: .github/workflows/fourpointoh-protector.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+              exit 0
+            else 
+              echo "❌ Label 'must-go-in-4.0' is absent"
```
Comment (mrnugget) [inline .github/workflows/fourpointoh-protector.yml:25]:
```suggestion
              echo "❌ Label 'i-acknowledge-this-goes-into-4.0' is absent"
```

**Comment by mrnugget** [comment]
> Run with dday set to 2022-09-09 and no label:

The last entry in your list -- shouldn't this be a different date?

**Review by mrnugget — CHANGES_REQUESTED** [review]
Thank you for doing this! I request changes because I think we should make clear with the label name that this is only an "ack" and not "please put this in the release", since authors of unrelated changes will also have to use it for their PRs.

Can we also make the "failed" message a bit more explanatory?

**Comment by jhchabran** [comment]
> Thank you for doing this! I request changes because I think we should make clear with the label name that this is only an "ack" and not "please put this in the release", since authors of unrelated changes will also have to use it for their PRs.
> 
> Can we also make the "failed" message a bit more explanatory?

It totally makes sense, thanks for suggesting this.

File: .github/workflows/fourpointoh-protector.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+          today=$(date +'%Y%m%d%H%M')
+          dday="202209090000"
+
+          if [ "$today" -gt "$dday" ]; then
+            if [ "$has_label" = "true" ]; then
+              echo "✅ Label 'i-acknowledge-this-goes-into-4.0' is present"
+              exit 0
+            else 
+              echo "❌ Label 'i-acknowledge-this-goes-into-4.0' is absent"
+              echo "👉 If you are 100% sure your changes should go in 4.0, label your PR with 'i-acknowledge-this-goes-into-4.0''"
```
Comment (mrnugget) [inline .github/workflows/fourpointoh-protector.yml:26]:
```suggestion
              echo "👉 We're in the Sourcegraph 4.0 code freeze. If you are 100% sure your changes should go in 4.0 or provide no risk to the release 4.0, add the label your PR with 'i-acknowledge-this-goes-into-4.0''"
```

**Review by mrnugget — APPROVED** [review]
Sweet!

**Comment by jhchabran** [comment]
> > Run with dday set to 2022-09-09 and no label:
> 
> The last entry in your list -- shouldn't this be a different date?

@mrnugget Why? We're launching on sep 9th 2022 right? I tested that before that this PR is not blocking anyone.

**Comment by mrnugget** [comment]
> Why? We're launching on sep 9th 2022 right? I tested that before that this PR is not blocking anyone.

Does that mean you have case 1 and 3 mixed up? In the year 202**1** a PR should not have a red check, right? Only >= 202**2**-09-09 and without a label, right?

screenshot_2022-08-29_16 32 09@2x

**Comment by jhchabran** [comment]
@mrnugget on the first scenario, the release day is set in the past, so you have to set a label for the check to pass. That particular test was "without" the label. It's basically the same thing as opening a PR on Sep 12th without the label, the check will block the PR. 

The third one, is when the release day is set on the Sep 9th (next week), but the PR being open today, it's passing without the label. 

I think the part you're missing is that dday is the release date, I just moved it in the past to check things. I didn't change the current date.

**Comment by mrnugget** [comment]
> I think the part you're missing is that dday is the release date, I just moved it in the past to check things. I didn't change the current date.

Oh, that was it. Thanks!

**Pull request merged**